### ISSUE TITLE
Chapter 7 - Out-of-Bag Evaluation

### ISSUE THREAD

**Issue by nicholashub** [ageron/handson-ml2#28]
In Chapter 7, the book says,

> The BaggingClassifier automatically performs soft voting instead of hard voting if the base classifier can estimate class probabilities (i.e., if it has a predict_proba() method), which is the case with Decision Trees classifiers

Then in the following section of Out-of-Bag Evaluation, it says the oob error is calculated by

> You can evaluate the ensemble itself by averaging out the oob evaluations of each predictor

It seems to me that this is just the average of generalization error for each tree. Why can it be used as the generalization error for the ensemble model, not to mention the random forest is trained using soft voting?

**Comment by ageron** [comment]
Hi @nicholashub ,

Thanks for your great feedback. Indeed, the second sentence could be much clearer. I'll fix it in the book.

What happens ([in this code](https://github.com/scikit-learn/scikit-learn/blob/78b7a50f23ce0710db196521884d86cc303bc182/sklearn/ensemble/_forest.py#L506) for random forest classifiers, and in [this code](https://github.com/scikit-learn/scikit-learn/blob/438bb3a438a42e23e3111c58575cf16841bb694d/sklearn/ensemble/_bagging.py#L611) for bagging classifiers) is that a prediction is made for each instance by using all the estimators that contain it in their oob set (appropriately using soft voting if all estimators have a `predict_proba()` method, or hard voting otherwise), and the `oob_score_` is computed based on these predictions (using the accuracy metric, i.e., the ratio of correct predictions).

Hope this helps. Closing this issue (feel free to reopen it if it's still unclear).
Thanks again!
